### ISSUE TITLE
Failed at saving my model correctly. It is saved but it's like the model had never been trained before.

### ISSUE THREAD

**Issue by Houdakilani996** [keras-team/keras#14000]
<em>Please make sure that this is a Bug or a Feature Request and provide all applicable information asked by the template.
If your issue is an **implementation question**, please ask your question on [StackOverflow](http://stackoverflow.com/questions/tagged/keras) or [on the Keras Slack channel](https://keras-slack-autojoin.herokuapp.com/) instead of opening a GitHub issue.</em>  

**System information**  
- Have I written custom code (as opposed to using example directory):  
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):  
- TensorFlow backend (yes / no):  
- TensorFlow version:  
- Keras version:  
- Python version:  
- CUDA/cuDNN version:  
- GPU model and memory:  

You can obtain the TensorFlow version with:  
python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"  
You can obtain the Keras version with:  
python -c 'import keras as k; print(k.__version__)'  

**Describe the current behavior**  

**Describe the expected behavior**  

**Code to reproduce the issue**  
Provide a reproducible test case that is the bare minimum necessary to generate the problem.  

**Other info / logs**  
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by Houdakilani996** [comment]
This is the error it gives me when i try to save it.
It is a warring, so the file is saved but there are some weights which are not saved i think : 
layer lstm_2 was passed non-serializable keyword arguments: {'initial_state': [<tf.tensor 's0:0' shape=(?, 256) dtype=float32>, <tf.tensor 'c0:0' shape=(?, 256) dtype=float32>]}. they will not be included in the serialized model (and thus will be missing at deserialization time).'. they will not be included '

**Comment by Houdakilani996** [comment]
THIS is my code : 
# -*- coding: utf-8 -*-
"""
Created on Tue Apr  7 11:29:28 2020

@author: KILANH
"""



from __future__ import print_function, division
from builtins import range
import pickle
import os
import string
import re
import numpy as np
import matplotlib.pyplot as plt
from pickle import dump
from unicodedata import normalize
from numpy import array
from keras.models import Model
from keras.layers import Input, LSTM, Dense, Embedding, \
  Bidirectional, RepeatVector, Concatenate, Dot, Lambda
from keras.preprocessing.text import Tokenizer
from keras.preprocessing.sequence import pad_sequences
import keras.backend as K
from keras.callbacks import ModelCheckpoint

from keras import optimizers
from pickle import load
from numpy.random import shuffle


# config
BATCH_SIZE = 64
EPOCHS = 100
LATENT_DIM = 256
LATENT_DIM_DECODER = 256 
EMBEDDING_DIM = 200


import tensorflow as tf
sess = tf.InteractiveSession()
with tf.Session() as sess:
     sess.run(tf.global_variables_initializer())
     
     
def load_doc(file):

	file = open(file, mode='rt', encoding='utf-8')
	text = file.read()
	file.close()
	return text

def to_pairs(doc):
	lines = doc.strip().split('\n')
	pairs = [line.split('\t') for line in  lines]
	return pairs

        
def clean_pairs(lines):
	cleaned = list()
	re_print = re.compile('[^%s]' % re.escape(string.printable))
	table = str.maketrans('', '', string.punctuation)
	for pair in lines:
		clean_pair = list()
		for line in pair:
			line = normalize('NFD', line).encode('ascii', 'ignore')
			line = line.decode('UTF-8')
			line = line.split()
			line = [word.lower() for word in line]
			line = [word.translate(table) for word in line]
			line = [re_print.sub('', w) for w in line]
			line = [word for word in line if word.isalpha()]
			clean_pair.append(' '.join(line))
		cleaned.append(clean_pair)
	return array(cleaned)

def save_clean_data(sentences, filename):
	dump(sentences, open(filename, 'wb'))
	print('Saved: %s' % filename)
    
def load_clean_sentences(filename):
    return load(open(filename, 'rb'))
    
filename = 'deu.txt'
doc = load_doc(filename)
pairs = to_pairs(doc)
shuffle(pairs)
clean_pairs = clean_pairs(pairs)
save_clean_data(clean_pairs, 'english-german.pkl')
raw_dataset = load_clean_sentences('english-german.pkl')
n_sentences = 20000
dataset = raw_dataset[:n_sentences, :]

train, test = dataset[:15000], dataset[15000:]

save_clean_data(dataset, 'english-german-both.pkl')
save_clean_data(train, 'english-german-train.pkl')
save_clean_data(test, 'english-german-test.pkl')


input_texts = [] # English sentences
target_texts = [] # German sentences
target_texts_inputs = []# english sentences offset by 1


#Adding <sos> and <eos> tokens

for lines in range(len(train)):
    target_texts_inputs.append('<sos>'+" "+ train[lines,1])
    
for lines in range(len(train)):
    target_texts.append(train[lines,1]+ " " +'<eos>')
    
for lines in range(len(train)):
    input_texts.append(train[lines,0]+"")


# tokenize English sentences
tokenizer_inputs = Tokenizer()
tokenizer_inputs.fit_on_texts(input_texts)
input_sequences = tokenizer_inputs.texts_to_sequences(input_texts)

# word to index mapping for English
word2idx_inputs = tokenizer_inputs.word_index

# Max length of English sentence
max_len_input = max(len(s) for s in input_sequences)

#tokenize GERMAN sentences
tokenizer_outputs = Tokenizer(filters='')
tokenizer_outputs.fit_on_texts(target_texts + target_texts_inputs) 
target_sequences = tokenizer_outputs.texts_to_sequences(target_texts)
target_sequences_inputs = tokenizer_outputs.texts_to_sequences(target_texts_inputs)

# Word to index mapping for GERMAN
word2idx_outputs = tokenizer_outputs.word_index

# store number of output words for later
num_words_output = len(word2idx_outputs) + 1

# Max length of GERMAN sentence
max_len_target = max(len(s) for s in target_sequences)

# pad the sequences
encoder_inputs = pad_sequences(input_sequences, maxlen=max_len_input)
decoder_inputs = pad_sequences(target_sequences_inputs, maxlen=max_len_target, padding='post')
decoder_targets = pad_sequences(target_sequences, maxlen=max_len_target, padding='post')

print('Data Preprocessed')
print("Max train German sentence length:", max_len_target)
print("Max train English sentence length:", max_len_input)
print("Unique English Vocabulary count:",len(word2idx_inputs))
print("Unique German Vocabulary count:",len(word2idx_outputs))


##Test Preparation
input_texts_test = [] # English sentences
target_texts_test = [] # German sentences
target_texts_inputs_test = []# German sentences offset by 1 for teacher forcing



#Adding <sos> and <eos> tokens

for lines in range(len(test)):
    target_texts_inputs_test.append('<sos>'+" "+ test[lines,1])
    
for lines in range(len(test)):
    target_texts_test.append(test[lines,1]+ " " +'<eos>')
    
for lines in range(len(test)):
    input_texts_test.append(test[lines,0])
    
    

# tokenize English sentences  for the test part :

input_sequences_test = tokenizer_inputs.texts_to_sequences(input_texts_test)


target_sequences_test = tokenizer_outputs.texts_to_sequences(target_texts_test)
target_sequences_inputs_test = tokenizer_outputs.texts_to_sequences(target_texts_inputs_test)

# pad the sequences
encoder_inputs_test = pad_sequences(input_sequences_test, maxlen=max_len_input)
decoder_inputs_test = pad_sequences(target_sequences_inputs_test, maxlen=max_len_target, padding='post')
decoder_targets_test = pad_sequences(target_sequences_test, maxlen=max_len_target, padding='post')

print("Max test German sentence length:", decoder_inputs_test.shape[1])
print("Max test English sentence length:", encoder_inputs_test.shape[1])



#Load pretrained word vectors from Glove
print('Loading word vectors...')
word2vec = {}
with open(os.path.join('glove.6B.%sd.txt' % EMBEDDING_DIM), encoding="utf8") as f:
  # is just a space-separated text file in the format:
  # word vec[0] vec[1] vec[2] ...
  for line in f:
    values = line.split()
    word = values[0]
    vec = np.asarray(values[1:], dtype='float32')
    word2vec[word] = vec
print('Found %s word vectors.' % len(word2vec))

# prepare embedding matrix
print('Filling pre-trained embeddings...')
num_words=len(word2idx_inputs) + 1
embedding_matrix = np.zeros((num_words, EMBEDDING_DIM))
for word, i in word2idx_inputs.items():
  if i < num_words:
    embedding_vector = word2vec.get(word)
    if embedding_vector is not None:
      # words not found in embedding index will be all zeros.
      embedding_matrix[i] = embedding_vector
      
      
def softmax_over_time(x):
  assert(K.ndim(x) > 2)
  e = K.exp(x - K.max(x, axis=1, keepdims=True))
  s = K.sum(e, axis=1, keepdims=True)
  return e / s

# create embedding layer
embedding_layer = Embedding(
  num_words,
  EMBEDDING_DIM,
  weights=[embedding_matrix],
  input_length=max_len_input,
  # trainable=True
  )


decoder_targets_one_hot = np.zeros(
  (
    len(input_texts),
    max_len_target,
    num_words_output
  ),
  dtype='float32'
)

# assign the values
for i, d in enumerate(decoder_targets):
  for t, word in enumerate(d):
    decoder_targets_one_hot[i, t, word] = 1
    
#encoder
encoder_inputs_placeholder = Input(shape=(max_len_input,))
x = embedding_layer(encoder_inputs_placeholder)
encoder = Bidirectional(LSTM(
  LATENT_DIM,
  return_sequences=True, dropout=0.2
))
encoder_outputs = encoder(x)

#Decoder
decoder_inputs_placeholder = Input(shape=(max_len_target,))
decoder_embedding = Embedding(num_words_output, EMBEDDING_DIM)
decoder_inputs_x = decoder_embedding(decoder_inputs_placeholder)


#Attention
attn_repeat_layer = RepeatVector(max_len_input)
attn_concat_layer = Concatenate(axis=-1)
attn_dense1 = Dense(10, activation='tanh')
attn_dense2 = Dense(1, activation=softmax_over_time)

attn_dot = Dot(axes=1) # to perform the weighted sum of alpha[t] * h[t]



def one_step_attention(h, st_1):
  # h = h(1), ..., h(Tx), shape = (Tx, LATENT_DIM * 2)
  # st_1 = s(t-1), shape = (LATENT_DIM_DECODER,)
  # copy s(t-1) Tx times
  # now shape = (Tx, LATENT_DIM_DECODER)
  st_1 = attn_repeat_layer(st_1)
  # Concatenate all h(t)'s with s(t-1)
  # Now of shape (Tx, LATENT_DIM_DECODER + LATENT_DIM * 2)
  x = attn_concat_layer([h, st_1])
  # Neural net first layer
  x = attn_dense1(x)
  # Neural net second layer with special softmax over time
  alphas = attn_dense2(x)
  # "Dot" the alphas and the h's
  # Remember a.dot(b) = sum over a[t] * b[t]
  context = attn_dot([alphas, h])

  return context

# define the rest of the decoder (after attention)
decoder_lstm = LSTM(LATENT_DIM_DECODER, return_state=True)
decoder_dense = Dense(num_words_output, activation='softmax')

initial_s = Input(shape=(LATENT_DIM_DECODER,), name='s0')
initial_c = Input(shape=(LATENT_DIM_DECODER,), name='c0')
context_last_word_concat_layer = Concatenate(axis=2)



# s, c will be re-assigned in each iteration of the loop
s = initial_s
c = initial_c

# collect outputs in a list at first
outputs = []
for t in range(max_len_target): # Ty times
  # get the context using attention
  context = one_step_attention(encoder_outputs, s)

  # we need a different layer for each time step
  selector = Lambda(lambda x: x[:, t:t+1])
  xt = selector(decoder_inputs_x)
  
  # combine 
  decoder_lstm_input = context_last_word_concat_layer([context, xt])
  o, s, c = decoder_lstm(decoder_lstm_input, initial_state=[s, c])
  # final dense layer to get next word prediction
  decoder_outputs = decoder_dense(o)
  outputs.append(decoder_outputs)
  
def stack_and_transpose(x):
  # x is a list of length T, each element is a batch_size x output_vocab_size tensor
  x = K.stack(x) # is now T x batch_size x output_vocab_size tensor
  x = K.permute_dimensions(x, pattern=(1, 0, 2)) # is now batch_size x T x output_vocab_size
  return x

# make it a layerx``
stacker = Lambda(stack_and_transpose)
outputs = stacker(outputs)

# create the model
model = Model(
  inputs=[
    encoder_inputs_placeholder,
    decoder_inputs_placeholder,
    initial_s, 
    initial_c,
  ],
  outputs=outputs
)

learning_rate=0.001
# checkpoint
filepath="english/german_finalmodel"
checkpoint = ModelCheckpoint(filepath, monitor='val_accuracy', verbose=1, save_best_only=True, mode='max')
callbacks_list = [checkpoint]

# compile the model

model.compile(optimizer=optimizers.Adam(learning_rate) ,loss='categorical_crossentropy', metrics=['accuracy'])

from keras.backend import manual_variable_initialization
manual_variable_initialization(True)


# train the model
z = np.zeros((encoder_inputs.shape[0], LATENT_DIM_DECODER)) # initial [s, c]
r = model.fit([encoder_inputs, decoder_inputs, z, z], decoder_targets_one_hot,batch_size=BATCH_SIZE,epochs=EPOCHS,validation_split=0.2,
    callbacks=callbacks_list)


# plot some data
plt.plot(r.history['loss'], label='loss')
plt.plot(r.history['val_loss'], label='val_loss')
plt.legend()
plt.show()

# accuracies
plt.plot(r.history['acc'], label='acc')
plt.plot(r.history['val_acc'], label='val_acc')
plt.legend()
plt.show()

encoder_model = Model(encoder_inputs_placeholder, encoder_outputs)

# next we define a T=1 decoder model
encoder_outputs_as_input = Input(shape=(max_len_input, LATENT_DIM * 2,))
decoder_inputs_single = Input(shape=(1,))
decoder_inputs_single_x = decoder_embedding(decoder_inputs_single)

# no need to loop over attention steps this time because there is only one step
context = one_step_attention(encoder_outputs_as_input, initial_s)

# combine context with last word
decoder_lstm_input = context_last_word_concat_layer([context, decoder_inputs_single_x])

# lstm and final dense
o, s, c = decoder_lstm(decoder_lstm_input, initial_state=[initial_s, initial_c])
decoder_outputs = decoder_dense(o)

# create the model object
decoder_model = Model(
  inputs=[
    decoder_inputs_single,
    encoder_outputs_as_input,
    initial_s, 
    initial_c
  ],
  outputs=[decoder_outputs, s, c]
)

decoder_model.save("enger_decoder_model.h5")
encoder_model.save("enger_encoder_model.h5")

from tensorflow.keras.models import load_model


#Loading the pre trained models:
encoder_model=load_model('enger_encoder_model.h5',compile=False)
decoder_model = load_model('enger_decoder_model.h5',compile=False, custom_objects={'softmax_over_time':softmax_over_time })
idx2word_eng = {v:k for k, v in word2idx_inputs.items()}
idx2word_trans = {v:k for k, v in word2idx_outputs.items()}


pickle.dump( idx2word_eng, open( "idx2word_eng1.pkl", "wb" ) )
pickle.dump(idx2word_trans,open("idx2word_germ.pkl","wb"))
pickle.dump(tokenizer_inputs,open("tokenizer_inputs.pkl","wb"))

def load_pkl(filename):
    return load(open(filename,'rb'))
idx2word_trans=load_pkl("idx2word_germ.pkl")

def decode_sequence(input_seq):
  # Encode the input as state vectors.
  enc_out = encoder_model.predict(input_seq)

  # Generate empty target sequence of length 1.
  target_seq = np.zeros((1, 1))
  
  # Populate the first character of target sequence with the start character.
  # NOTE: tokenizer lower-cases all words
  target_seq[0, 0] = word2idx_outputs['<sos>']

  # if we get this we break
  eos = word2idx_outputs['<eos>']


  # [s, c] will be updated in each loop iteration
  s = np.zeros((1, LATENT_DIM_DECODER))
  c = np.zeros((1, LATENT_DIM_DECODER))


  # Create the translation
  output_sentence = []
  for _ in range(max_len_target):
    o, s, c = decoder_model.predict([target_seq, enc_out, s, c])
        

    # Get next word
    idx = np.argmax(o.flatten())

    # End sentence of EOS
    if eos == idx:
      break

    word = ''
    if idx > 0:
      word = idx2word_trans[idx]
      output_sentence.append(word)

    # Update the decoder input
    # which is just the word just generated
    target_seq[0, 0] = idx

  return ' '.join(output_sentence)
test_actual_sentence=[]
test_predicted_sentence=[]

sess = tf.Session()
tf.initialize_all_variables().run()
    
for i in range(len(test)):
    # Do some test translations
  
  input_seq = encoder_inputs_test[i:i+1]
  translation = decode_sequence(input_seq)

  test_actual_sentence.append(target_texts_test[i])
  test_predicted_sentence.append(translation)

  
for i in np.random.randint(0,1400,10):
    print('-')
    print('Input sentence:', input_texts_test[i])
    print('Predicted translation:', test_predicted_sentence[i])
    print('Actual translation:', target_texts_test[i])
    





from langdetect import detect


def detectLanguage(sentence):
    res=detect(sentence)
    if res =="en":
        print("This sentence is in:",res)
        print("We will need to translate it in German")
        sentence=[sentence]
        tokenizer_inputs.fit_on_texts(sentence)
        input_s= tokenizer_inputs.texts_to_sequences(sentence)
        enc = pad_sequences(input_s, maxlen=max_len_input)
        translation = decode_sequence(enc)
        print('Result: ' + translation)      
    elif res =="fr":
        print("we will translate this sentence from french to german")
    

example = input("translate this sentence: ")
Res=detectLanguage(example)
print(detectLanguage(example))

# without detecting :
sentence = input("translate this sentence: ")
sentence=[sentence]
tokenizer_inputs.fit_on_texts(sentence)
input_s= tokenizer_inputs.texts_to_sequences(sentence)
enc = pad_sequences(input_s, maxlen=max_len_input)
translation = decode_sequence(enc)
print('Result: ' + translation)

**Comment by jvishnuvardhan** [comment]
@Houdakilani996 This is a stale issue. 

Is this still an issue for you? Can you please share a simple standalone code to reproduce the issue? Also, try running the code with `TF2.6` or `tf-nightly` and let us know whether the issue persist with recent TF versions. Thanks!

**Comment by jvishnuvardhan** [comment]
Closing due to lack of recent activity. Please feel free to reopen when you have more information. Thanks!
